K8s: add k8s container and node (host) discovery

hunchback opened the pull request:
This is a 2nd phase for the PR adding this capability for node discovery via the k8s probe itself (without relaying on the host agent).

To test just connect to a k8s landscape scape

lebauce commented:
When the 'docker' probe is used, nodes of type 'container' are already created. They are added as child of the namespace node.

In case where the node do not run the Skydive agent, we have no "namespace" node. I'm not sure what to do but we should have a consistent way of representing things, whether we run the Skydive agent or not.

hunchback commented:
@lebauce - the docker will add container to the namespace node, but if the k8s detects a container is under a pod it will place the container under the pod.

hunchback commented:
@lebauce / @safchain  - please view this as yet another intermediate step in k8s probe implementation, which currently provides (compared to the current state):
- ability to view k8s landscape (including hosts and containers) even w/o an agent probe
- merge to metadata in the case of hosts, containers when agent has created these already

Known limitations:
- missing periodic refresh of all the k8s probe created graph elements (configurable with a default of 30s)
- we must mark a host or container created by k8s in a color indicating its' not an accessible element (e.g. such as ability to inject/capture traffic etc.) and block a-priori any attempt to run any skydive-agent related operations on these nodes.
- we must map a hosts, containers to the same TID regardless of if it was created by the agent or the analyzer/k8s probe
- we must have a clear separation of  data/metadata added by the agent and data/metadata added by the analyzer, we could either have separate graph-nodes or have a clear separation within the graph-node element ... 
- we must support views with only k8s; only agent; merged view in the UI
- the current HOST DELETE message sent to the analyzer will also remove the k8s probe data (for now its' lost until some event occurs), we must change so that HOST DELETE only deletes the data/metadata added by the requested client (such that if issued by the agent, only data/metadata added by the agent will be deleted by this request, and if k8s added data to the host view, these graph-elements will not be deleted)

safchain commented:
@hunchback Thanks, It's fine for me to have intermediate steps. 

For the limitations, we really need to discuss about what to do with nodes created in two places and if we want to support this kind of deployment as it as deep design impact. 

For the host delete stuff, playing with the host binding (host field) could help. 

Agree that we should mark nodes reported only from the controller/config side. Skydive was designed to not rely on controller/config and to validate what is really done (namespaces, etc), so need to distinguish both ways of getting the things.

That said I'll review it asap

safchain commented on etc/skydive.yml.default:
```diff
@@ -253,5 +253,11 @@ ui:
+  # The kube config file, if executed inside the cluster, you may use the empty
+  # file, kube-config-empty.yml
   # config_file: /etc/skydive/kube-cfg.yml
+  subprobes:
```
Maybe we should put something in order to say that is experimental.

hunchback replied:
@safchain - ok have added a comment in the above configuration file

safchain commented on topology/probes/k8s/container.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+func makeContainerUID(pod *v1.Pod, containerName string) string {
+	return string(pod.GetUID()) + "-" + containerName
```
We should use a uuidv5 with pod UID and containerName instead of concatenating so that we will keep the UUID of graph element consistent in term of format.

hunchback replied:
ok, have added a new function graph.NewIDNameBased(namespace, name string) which is now used to create a well-formed UID.

safchain commented on topology/probes/k8s/container.go:
```diff
@@ -0,0 +1,171 @@
+		containerNode = c.graph.NewNode(graph.Identifier(uid), c.newMetadata(pod, container))
+	} else {
+		logging.GetLogger().Infof("container{%s} already exists", container.Name)
+		addMetadata(c.graph, containerNode, container)
```
I'm sure we should copy all the container metadata directly since we are going to move more or less to something schema based, but I think that is acceptable for now.

safchain commented on topology/probes/k8s/graph.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+func addMetadata(g *graph.Graph, n *graph.Node, extra interface{}) {
+	g.AddMetadata(n, "Manager", "k8s")
```
It would be better to use a transaction otherwise it will generate 2 updates and 2 graph revisions

hunchback replied:
ok, using transactions

safchain commented on topology/probes/k8s/graph.go:
```diff
@@ -0,0 +1,42 @@
+
+func addMetadata(g *graph.Graph, n *graph.Node, extra interface{}) {
+	g.AddMetadata(n, "Manager", "k8s")
+	g.AddMetadata(n, "K8s", extra)
```
Same here, not that we should put everything directly, but acceptable for now I guess

hunchback replied:
yes, using transactions now

safchain commented on topology/probes/k8s/k8s.go:
```diff
@@ -23,49 +23,87 @@
-func (k8s *probe) Start() {
-	k8s.networkPolicyCache.Start()
-	k8s.podCache.Start()
+type starter interface {
```
there is already a probe interface probe/bundle.go

hunchback replied:
replaced with probe.Probe

safchain commented on topology/probes/k8s/k8s.go:
```diff
@@ -23,49 +23,87 @@
-func (k8s *probe) Stop() {
-	k8s.networkPolicyCache.Stop()
-	k8s.podCache.Stop()
+func (probe *Probe) getSubProbes() []starter {
```
same here you could use the proble/bundle.go

hunchback replied:
Replaced map of subProbes with probe.ProbeBundle

hunchback commented:
@lebauce - I have added a set of commits addressing all the above comments, for easing review of newly added code and did not squash commits together, once reviewed I will do so (kindly indicate which you would like to merge together).

lebauce commented:
It would be nice to enable all or some subprobes by default:

```
diff --git a/config/config.go b/config/config.go
index c1af3b1a..221d0bdc 100644
--- a/config/config.go
+++ b/config/config.go
@@ -100,6 +100,9 @@ func init() {
 
        cfg.SetDefault("host_id", host)
 
+       cfg.SetDefault("k8s.config_file", "/etc/skydive/kube-cfg.yml")
+       cfg.SetDefault("k8s.subprobes", []string{"networkpolicy", "pod", "container", "node"})
+
        cfg.SetDefault("logging.backends", []string{"stderr"})
        cfg.SetDefault("logging.color", true)
        cfg.SetDefault("logging.encoder", "")
@@ -116,8 +119,6 @@ func init() {
        cfg.SetDefault("sflow.port_min", 6345)
        cfg.SetDefault("sflow.port_max", 6355)
 
-       cfg.SetDefault("k8s.config_file", "/etc/skydive/kube-cfg.yml")
-
        cfg.SetDefault("storage.elasticsearch.host", "127.0.0.1:9200")
        cfg.SetDefault("storage.elasticsearch.maxconns", 10)
        cfg.SetDefault("storage.elasticsearch.retry", 60)
```

hunchback commented:
```
With the elasticsearch backend, I get "Bulk Insertion" errors. That's because elasticsearch does not accept dots in the keys of maps. In my case, it happens for the key "alpha.kubernetes.io/provided-node-ip" in Node.metadata.annotations.

In the Docker probe, we do https://github.com/skydive-project/skydive/blob/master/topology/probes/docker.go#L116-L122
```

I don't really like the approach used for Docker:
(1) it alters the structure `m["a.b.c/d"] ="e"` is now m["a"]["b"]["c/d"] = "e"`
(2) it is non-generic and requires handling each potentially offending field specifically

I have altered `common.NormalizeValue()`:
(1) it will convert `m["a.b.c/d"] = "e"` to `m["a_b_c/d"] = "e"`
(2) it traverses the handed object and make corrections recursively

hunchback commented:
> If the agent is started agent on the node that runs kubelet, there will be 2 nodes of type "host". We discussed with @safchain and thought it may indeed be simpler to create dedicated nodes of type "node" (or "k8s-node", or something else) and attach the pods to it. And we could simply create links between the hosts and the nodes when both of them has showed up. This way, we don't have problem with merging metadata and we make a clean distinction between the logical view and the physical view. What do you think ?

@lebauce - as you proposed creating a dedicated graph.Node representing the k8s node element (of Type: "node"), such that `pods --owned--> node --owned--> host`

hunchback commented on etc/skydive.yml.default:
```diff
@@ -253,5 +253,12 @@ ui:
 
 k8s:
-  # The KUBECONFIG yml file, if k8sctl is executed inside the cluster, you may use the empty config file - kube-config-empty.yml
+  # EXPERIMENTAL: k8s probe still under development and should be used in productive systems
```
ok, handled

hunchback commented on topology/graph/graph.go:
```diff
@@ -269,6 +269,15 @@ func GenID() Identifier {
+// GenIDNameBased helper generate a node Identifier
+func GenIDNameBased(namespace, name string) Identifier {
+	ns, _ := uuid.ParseHex(namespace)
+	n := []byte(name)
```
ok

hunchback commented on topology/probes/k8s/container.go:
```diff
@@ -0,0 +1,170 @@
+
+func (c *containerCache) newMetadata(pod *v1.Pod, container *v1.Container) graph.Metadata {
+	m := newMetadata("container", container.Name, container)
+	common.SetField(m, DockerNameField, container.Name)
```
ok - added `SetField(k, v)`

hunchback commented on topology/probes/k8s/graph.go:
```diff
@@ -0,0 +1,44 @@
+
+func addMetadata(g *graph.Graph, n *graph.Node, extra interface{}) {
+	tr := g.StartMetadataTransaction(n)
+	defer tr.Commit()
```
ok

hunchback commented on topology/probes/k8s/pod.go:
```diff
@@ -41,100 +40,119 @@ type podCache struct {
-		"ObjectMeta": pod.ObjectMeta,
-		"Spec":       pod.Spec,
-	}
+func newPodIndex(g *graph.Graph, by string) *graph.MetadataIndexer {
```
ok, renamed to `newPodIndexer`

hunchback commented:
> It would be nice to enable all or some subprobes by default:

added

hunchback commented:
I have addressed all above comments in the following commits:

```
Applying: build: fix dependency on 'dlv' for 'make debug'
Applying: k8s: normalized keys in 'K8s' to ES friendly form
Applying: k8s: create a dedicate k8s:node grpah.Node
Applying: k8s: fixed configuration *EXPERIMENTAL* message
Applying: k8s: minor fix in graph.GenIDNameBased()
Applying: k8s: added graph.Metadata.SetField(k, v)
Applying: k8s: removed defer in k8s.addMetadata()
Applying: k8s: replaced k8s.newPodIndex() -> k8s.newPodIndexer()
Applying: k8s: added default configuration values
```

Once reviewed I can squash these commits with existing ones

lebauce commented:
@hunchback Regarding NormalizeValue, I agree it's not optimal to alter the structure. What you suggest is what we did when we first faced this bug. It caused other problems, in gremlin parsing for instance (g.V().HasKey('Labels.My.Label.with.a.dot') should be written g.V().HasKey('Labels["My.Label.with.a.dot"]). So I suggest you replace dots with underscores in the "k8s" probe only, and not in common/types.go

Otherwise it looks good to me. With the NormalizeValue change, I'll happily merge it.

Maybe you can squash the following commits into one commit:

k8s: minor cleanup of configuration
k8s: add a comment that k8s probe is experimental
k8s: added default configuration values
k8s: fixed configuration *EXPERIMENTAL* message

hunchback commented:
> @hunchback Regarding NormalizeValue, I agree it's not optimal to alter the structure. What you suggest is what we did when we first faced this bug. It caused other problems, in gremlin parsing for instance (g.V().HasKey('Labels.My.Label.with.a.dot') should be written g.V().HasKey('Labels["My.Label.with.a.dot"]). So I suggest you replace dots with underscores in the "k8s" probe only, and not in common/types.go

@lebauce - I have changed `common.NormalizeValue()` to split keys by using `SetField()` - so all including k8s probe will use the dot-notation not the under_score_delimited, in a separate PR will can then remove the specific Docker code (now redundant).

> k8s: minor cleanup of configuration
> k8s: add a comment that k8s probe is experimental
> k8s: added default configuration values
> k8s: fixed configuration EXPERIMENTAL message

Have been merged into: 
- k8s: config cleanup and default values for probe

lebauce commented on vendor/vendor.json:
```diff
@@ -1935,6 +1941,10 @@
 			"path": "gopkg.in/yaml.v2",
 			"revision": "bef53efd0c76e49e6de55ead051f886bea7e9420"
 		},
+		{
```
nit: This can be removed

hunchback replied:
Hi @lebauce when removing it I get the below, but regardless, I would be happy to address (if needed) in a separate PR, as I need the k8s in the community version to support our internal integration efforts ....

```
make[1]: Entering directory '/home/vagrant/projects/gopath/src/github.com/skydive-project/skydive/contrib/snort'
govendor build snortSkydive.go
../../vendor/github.com/spf13/viper/util.go:29:2: cannot find package "gopkg.in/yaml.v2" in any of:
        /home/vagrant/projects/gopath/src/github.com/skydive-project/skydive/vendor/gopkg.in/yaml.v2 (vendor tree)
        /usr/local/go/src/gopkg.in/yaml.v2 (from $GOROOT)
        /home/vagrant/projects/gopath/src/gopkg.in/yaml.v2 (from $GOPATH)
Error: exit status 1
Makefile:6: recipe for target 'snortSkydive' failed
```
